exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/TypeInferenceUtils.java

```diff
@@ -0,0 +1,568 @@
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlCharStringLiteral;
+import org.apache.calcite.sql.SqlDynamicParam;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperatorBinding;
+import org.apache.calcite.sql.type.SqlReturnTypeInference;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import org.apache.drill.common.expression.ExpressionPosition;
+import org.apache.drill.common.expression.FunctionCall;
+import org.apache.drill.common.expression.LogicalExpression;
+import org.apache.drill.common.expression.MajorTypeInLogicalExpression;
+import org.apache.drill.common.exceptions.UserException;
+import org.apache.drill.common.types.TypeProtos;
+import org.apache.drill.common.types.Types;
+import org.apache.drill.exec.expr.TypeHelper;
+import org.apache.drill.exec.expr.fn.DrillFuncHolder;
+import org.apache.drill.exec.planner.logical.DrillConstExecutor;
+import org.apache.drill.exec.resolver.FunctionResolver;
+import org.apache.drill.exec.resolver.FunctionResolverFactory;
+import org.apache.drill.exec.resolver.TypeCastRules;
+
+import java.util.List;
+import java.util.Map;
+
+public class TypeInferenceUtils {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(TypeInferenceUtils.class);
+
+  public static final TypeProtos.MajorType UNKNOWN_TYPE = TypeProtos.MajorType.getDefaultInstance();
+  private static ImmutableMap<TypeProtos.MinorType, SqlTypeName> DRILL_TO_CALCITE_TYPE_MAPPING =
+      ImmutableMap.<TypeProtos.MinorType, SqlTypeName> builder()
+          .put(TypeProtos.MinorType.INT, SqlTypeName.INTEGER)
+          .put(TypeProtos.MinorType.BIGINT, SqlTypeName.BIGINT)
+          .put(TypeProtos.MinorType.FLOAT4, SqlTypeName.FLOAT)
+          .put(TypeProtos.MinorType.FLOAT8, SqlTypeName.DOUBLE)
+          .put(TypeProtos.MinorType.VARCHAR, SqlTypeName.VARCHAR)
+          .put(TypeProtos.MinorType.BIT, SqlTypeName.BOOLEAN)
+          .put(TypeProtos.MinorType.DATE, SqlTypeName.DATE)
+          .put(TypeProtos.MinorType.DECIMAL9, SqlTypeName.DECIMAL)
+          .put(TypeProtos.MinorType.DECIMAL18, SqlTypeName.DECIMAL)
+          .put(TypeProtos.MinorType.DECIMAL28SPARSE, SqlTypeName.DECIMAL)
+          .put(TypeProtos.MinorType.DECIMAL38SPARSE, SqlTypeName.DECIMAL)
+          .put(TypeProtos.MinorType.TIME, SqlTypeName.TIME)
+          .put(TypeProtos.MinorType.TIMESTAMP, SqlTypeName.TIMESTAMP)
+          .put(TypeProtos.MinorType.VARBINARY, SqlTypeName.VARBINARY)
+          .put(TypeProtos.MinorType.INTERVALYEAR, SqlTypeName.INTERVAL_YEAR_MONTH)
+          .put(TypeProtos.MinorType.INTERVALDAY, SqlTypeName.INTERVAL_DAY_TIME)
+          .put(TypeProtos.MinorType.MAP, SqlTypeName.MAP)
+          .put(TypeProtos.MinorType.LIST, SqlTypeName.ARRAY)
+          .put(TypeProtos.MinorType.LATE, SqlTypeName.ANY)
```

Comment:
I recollect there were more types here, albeit commented out. Can you add them back (including comments)?